### PR TITLE
add Keccak256 hash to cadence

### PR DESCRIPTION
**Pull request by zhiqiangxu**
This is a PR for https://github.com/onflow/cadence/issues/537 .

I'll make another PR to the prefixedHashing in flow-go repo once this is merged (in order to reference to the newly added `runtime.HashAlgorithmKECCAK_256`).

Basically the PR for flow-go will be this commit:

https://github.com/zhiqiangxu/flow-go/commit/317a09428d6defbc51e9eb3caa8ab2e74903e261

### CODE REVIEW

**Comment by turbolent** [comment]
Thank you for contributing @zhiqiangxu! This looks good 👍 

From what I can see, this is all that is needed to be changed. However, after adding the new "enum case" to `HashAlgorithm`, `hashalgorithm_string.go` must be updated. You can update it by running `make generate`.

**Comment by turbolent** [comment]
@tarakby Could you please have a look?

**Review by tarakby — COMMENTED** [review]
Thank you for the contribution! 
This looks good. I just left minor comments about the doc, but I let the Cadence team approve the rest of the Cadence related PR. 
Looking forward to the flow-go PR :)

**Review by ramtinms — APPROVED** [review]
Looks good to me

**Comment by turbolent** [comment]
@zhiqiangxu I investigated the CI test failures. The addition of the new enum case causes a different iteration order, so the assertions in the failing tests need to be updated. Is it OK to commit the fixes to your branch?

File: docs/language/crypto.mdx [hunk 1]
```diff
@@ -27,6 +27,9 @@ pub enum HashAlgorithm: UInt8 {
     /// used in BLS signatures.
     pub case KMAC128_BLS_BLS12_381 = 5
 
+    /// KECCAK_256 is ethereum-compatible Keccak-256 hash
```
[the remark this answers is not in the document]
Comment (dete) [reply docs/language/crypto.mdx]: Let's not remove Ethereum entirely from that description, perhaps? My suggestion:

```
/// KECCAK_256 is the 256-bit legacy Keccak algorithm (NIST SHA3 candidate submission), as used by Ethereum and others
```

[the remark this answers is not in the document]
Comment (zhiqiangxu) [reply docs/language/crypto.mdx]: Ethereum is not removed entirely, it's on the 2nd line:)

**Review by turbolent — APPROVED** [review]
Thank you for your contribution! 👍

**Comment by tarakby** [comment]
FYI, @ramtinms and I will be adding Keccak-256 to Flow-go next week, for both the crypto and FVM side. 
@zhiqiangxu, letting you know so we don't duplicate the work.

**Comment by zhiqiangxu** [comment]
> FYI, @ramtinms and I will be adding Keccak-256 to Flow-go next week, for both the crypto and FVM side. @zhiqiangxu, letting you know so we don't duplicate the work.

ok, but seems that CI is still not happy?

**Pull request merged**